x/gamm/keeper/migrate.go

```diff
@@ -109,27 +109,52 @@ func (k Keeper) GetLinkedBalancerPoolID(ctx sdk.Context, concentratedPoolId uint
 	return sdk.BigEndianToUint64(balancerPoolIdBigEndian), nil
 }
 
-// OverwriteMigrationRecords sets the balancer to gamm pool migration info to the store
-// Deletes all existing records, migrationInfo in state is completely overwitten by the given migrationInfo.
-func (k Keeper) OverwriteMigrationRecords(ctx sdk.Context, migrationInfo types.MigrationRecords) {
+// deleteMigrationKeys deletes all keys with the given prefixKey.
+func (k Keeper) deleteMigrationKeys(ctx sdk.Context, prefixKey []byte) {
 	store := ctx.KVStore(k.storeKey)
-
-	// delete all existing keys
-	prefixStore := prefix.NewStore(store, types.KeyPrefixMigrationInfoBalancerPool)
+	prefixStore := prefix.NewStore(store, prefixKey)
 	iter := prefixStore.Iterator(nil, nil)
 	defer iter.Close()
 
 	for ; iter.Valid(); iter.Next() {
 		prefixStore.Delete(iter.Key())
 	}
+}
 
-	prefixStore = prefix.NewStore(store, types.KeyPrefixMigrationInfoCLPool)
-	iter = prefixStore.Iterator(nil, nil)
-	defer iter.Close()
+// OverwriteMigrationRecords sets the balancer to gamm pool migration info to the store and deletes all existing records
+// migrationInfo in state is completely overwitten by the given migrationInfo.
+// Additionally, the distribution record for the balancer pool is modified to redirect incentives to the new concentrated pool.
+func (k Keeper) OverwriteMigrationRecords(ctx sdk.Context, migrationInfo types.MigrationRecords) error {
+	store := ctx.KVStore(k.storeKey)
 
-	for ; iter.Valid(); iter.Next() {
-		prefixStore.Delete(iter.Key())
+	// delete all existing keys
+	k.deleteMigrationKeys(ctx, types.KeyPrefixMigrationInfoBalancerPool)
+	k.deleteMigrationKeys(ctx, types.KeyPrefixMigrationInfoCLPool)
+
+	for _, balancerToCLPoolLink := range migrationInfo.BalancerToConcentratedPoolLinks {
+		balancerToClPoolKey := types.GetKeyPrefixMigrationInfoBalancerPool(balancerToCLPoolLink.BalancerPoolId)
+		store.Set(balancerToClPoolKey, sdk.Uint64ToBigEndian(balancerToCLPoolLink.ClPoolId))
+
+		clToBalancerPoolKey := types.GetKeyPrefixMigrationInfoPoolCLPool(balancerToCLPoolLink.ClPoolId)
+		store.Set(clToBalancerPoolKey, sdk.Uint64ToBigEndian(balancerToCLPoolLink.BalancerPoolId))
+
+		err := k.redirectDistributionRecord(ctx, balancerToCLPoolLink.BalancerPoolId, balancerToCLPoolLink.ClPoolId)
+		if err != nil {
+			return err
+		}
 	}
+	return nil
+}
+
+// OverwriteMigrationRecordsNoRedirect behaves the same as OverwriteMigrationRecords but does not modify the distribution record.
+// This is specifically used in init genesis, since we want to overwrite the migration records but not modify the distribution record,
+// since this should be handled at time of genesis export by the incentive module.
+func (k Keeper) OverwriteMigrationRecordsNoRedirect(ctx sdk.Context, migrationInfo types.MigrationRecords) {
```

Comment:
Why does this have to be overwritten? If this is meant to be used in InitGenesis, let's call this simple `SetMigrationRecords` and remove the deletion logic

Reply:
Yeah its only meant to be called in initGenesis, renamed and removed deletion logic here https://github.com/osmosis-labs/osmosis/pull/5400/commits/6a449dce9f2a6ffd8768d3394c47a7d4f9a3bbbd